File: Common/Src/ChinesePuzzle/Game/CardBoard.cpp
/**
 *  CardBoard.cpp
 *  ChinesePuzzle
 *
 *  Created by Mathieu LEDRU on 01/11/11.
 *
 *  GPL License:
 *  Copyright (c) 2011, Mathieu LEDRU
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include "cpMacro.h"
#include "CardBoard.h"
#include "GameConfig.h"

using namespace cocos2d;

CardBoard::CardBoard() :
emptySprite(NULL),
yesSprite(NULL),
noSprite(NULL),
state(CardBoardEmpty)
{
}

CardBoard::~CardBoard()
{
	CC_SAFE_RELEASE(emptySprite);
	CC_SAFE_RELEASE(yesSprite);
	CC_SAFE_RELEASE(noSprite);
}

CardBoard* CardBoard::cardBoardWithConf(GameConfigCommon* conf)
{
	CardBoard* cardBoard = new CardBoard();
	if (cardBoard && cardBoard->initCardBoardWithConf(conf))
	{
        cardBoard->autorelease();
        return cardBoard;
    }
    CC_SAFE_DELETE(cardBoard);
	return NULL;
}

bool CardBoard::initCardBoardWithConf(GameConfigCommon* conf)
{
	if(!CCSpriteBatchNode::initWithTexture(ccTextureNull, 1))
	{
		return false;
	}
	
	this->setConf(conf);
	
	return true;
}

CardBoardState CardBoard::getState()
{
	return this->state;
}

void CardBoard::setState(CardBoardState state)
{
	this->setState(state, false);
}

void CardBoard::setState(CardBoardState state, bool force)
{
	if(this->state != state || force)
	{
		switch (state)
		{
			case CardBoardEmpty: copySpriteBatchNode(emptySprite, this); break;
			case CardBoardYes: copySpriteBatchNode(yesSprite, this); break;
			case CardBoardNo: copySpriteBatchNode(noSprite, this); break;
		}
		
		this->state = state;
	}
}

void CardBoard::setConf(GameConfigCommon* conf)
{
	if(emptySprite == NULL)
	{
		emptySprite = CCSpriteBatchNode::batchNodeWithTexture(ccTextureNull);
		emptySprite->retain();
	}
	if(yesSprite == NULL)
	{
		yesSprite = CCSpriteBatchNode::batchNodeWithTexture(ccTextureNull);
		yesSprite->retain();
	}
	if(noSprite == NULL)
	{
		noSprite = CCSpriteBatchNode::batchNodeWithTexture(ccTextureNull);
		noSprite->retain();
	}
	conf->getNodeThemePath("cardboardempty", emptySprite);
	conf->getNodeThemePath("cardboardyes", yesSprite);
	conf->getNodeThemePath("cardboardno", noSprite);
	
	this->setState(this->getState(), true);
}

File: Common/Bin/Data/js/Game/CardBoard.js
/*
This file is part of the ChinesePuzzle package.

(c) Mathieu Ledru

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.
*/

cpz.CardBoardState = {
  Empty: 1,
  Yes: 2,
  No: 3
};

cpz.CardBoard = cpz.Card.extend({
  _emptySprite: null,
  _yesSprite: null,
  _noSprite: null,
  _state: null,
  ctor: function() {
    this._super();
    return this._state = cpz.CardBoardState.Empty;
  },
  initWithConf: function(conf) {
    if (!this.initWithTexture(cc.textureNull(), 1)) {
      return false;
    }
    this.setConf(conf);
    return true;
  },
  getState: function() {
    return this._state;
  },
  setState: function(state, force) {
    if (force == null) {
      force = false;
    }
    if (this._state !== state || force) {
      switch (state) {
        case cpz.CardBoardState.Yes:
          cc.copySpriteBatchNode(this._yesSprite, this._batchNode);
          break;
        case cpz.CardBoardState.No:
          cc.copySpriteBatchNode(this._noSprite, this._batchNode);
          break;
        default:
          cc.copySpriteBatchNode(this._emptySprite, this._batchNode);
      }
      return this._state = state;
    }
  },
  setConf: function(conf) {
    if (this._emptySprite === null) {
      this._emptySprite = cc.SpriteBatchNode.createWithTexture(cc.textureNull(), 1);
    }
    if (this._yesSprite === null) {
      this._yesSprite = cc.SpriteBatchNode.createWithTexture(cc.textureNull(), 1);
    }
    if (this._noSprite === null) {
      this._noSprite = cc.SpriteBatchNode.createWithTexture(cc.textureNull(), 1);
    }
    conf.getNodeThemePath('cardboardempty', this._emptySprite);
    conf.getNodeThemePath('cardboardyes', this._yesSprite);
    conf.getNodeThemePath('cardboardno', this._noSprite);
    return this.setState(this.getState, true);
  }
});

cpz.CardBoard.createWithConf = function(conf) {
  var obj;
  obj = new cpz.CardBoard();
  if (obj && obj.initWithConf(conf)) {
    return obj;
  }
  return null;
};
